pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -729,10 +727,16 @@ func (c *Controller) getProxyServiceInstancesByPod(pod *v1.Pod, service *v1.Serv
 		}
 	}
 
+	initEndpoint := c.newIstioEndpoint(pod)
 	for tp, svcPort := range tps {
 		// consider multiple IP scenarios
 		for _, ip := range proxy.IPAddresses {
-			out = append(out, c.getEndpoints(podIP, ip, int32(tp.Port), svcPort, svc))
+			istioEndpoint := c.completeIstioEndpoint(initEndpoint, ip, int32(tp.Port), svcPort.Name, svc)
```

Comment:
I may be missing some thing fundamental here, what do we gain by splitting the endpoint construction in to two stages?

Reply:
By this the first `initIstioEndpoint` will be called only once.

Reply:
I see. Is that a big deal, given there will be very few service instances? of course, if the number of services are huge, it might be useful. Curious, did you find this to be a problem in some envs or coming up in profiles or general observation?

Reply:
TBH, this is not a big deal now. But we have many services that exposes several ports. And i remember guys from rancher uses multiple ips per pod.